ulmfit/pretrain_lm.py

```diff
@@ -57,13 +54,26 @@ def pretrain_lm(dir_path, lang='en', cuda_id=0, qrnn=True, clean=True, max_vocab
     if qrnn:
         print('Using QRNNs...')
 
-    trn_path = dir_path / f'{lang}.wiki.train.tokens'
-    val_path = dir_path / f'{lang}.wiki.valid.tokens'
-    tst_path = dir_path / f'{lang}.wiki.test.tokens'
+    trn_path = dir_path / f'{lang}.wiki.train.tokens.unk'
+    val_path = dir_path / f'{lang}.wiki.valid.tokens.unk'
+    tst_path = dir_path / f'{lang}.wiki.test.tokens.unk'
     for path_ in [trn_path, val_path, tst_path]:
         assert path_.exists(), f'Error: {path_} does not exist.'
 
-    if clean:
+    if subword:
+        # apply sentencepiece tokenization
+        trn_path = dir_path / f'{lang}.wiki.train.tokens'
+        val_path = dir_path / f'{lang}.wiki.valid.tokens'
+
+        read_file(trn_path, 'train')
+        read_file(val_path, 'valid')
+
+        sp = get_sentencepiece(dir_path, trn_path, name)
+
+        data_lm = TextLMDataBunch.from_csv(dir_path, **sp)
+        itos = data_lm.train_ds.vocab.itos
```

Comment:
No, but it has a vocab and we should be able to pull these out from that (I think it already will). I'm waiting to test this end-to-end.

Reply:
`itos` might have same semantics for sp as for normal word tokenization as in both cases you convert integers to text repr. But `stoi` won't have similar meaning as you won't be able to tokenize the text to words and pass it to `stoi`.   How about implementing your own version of Vocab has the same methods but does not have itos or stoi?

Reply:
@PiotrCzapla I am not following your reasoning. `stoi` does have meaning here as well since SentencepieceTokenizer will output the list of string pieces that will exactly match the parts that were generated by the sp model and then dumped into the `itos.pkl` file. It is a little more inefficient to be aligned with the Tokenizer/Vocab API since in this way we are first mapping the sentences into a list of strings and only then mapping to integers, while the sentencepiece utility can actually do that in one go.

Reply:
@eisenjulian  my bad, I didn't notice how it is being done actually.  Indeed a quite clever way to adapt to the fastai API.  :)